BlueprintUI/Sources/Element/ElementContent.swift

```diff
@@ -154,30 +200,116 @@ extension ElementContent.Builder: ContentStorage {
         return result
     }
 
-    private var layoutItems: [(LayoutType.Traits, Measurable)] {
-        return children.map { ($0.traits, $0) }
+    private func layoutItems(in environment: Environment) -> [(LayoutType.Traits, Measurable)] {
+        return children.map { ($0.traits, $0.content.measurable(in: environment)) }
     }
 
 }
 
 
 extension ElementContent.Builder {
 
-    fileprivate struct Child: Measurable {
+    fileprivate struct Child {
 
         var traits: LayoutType.Traits
         var key: String?
         var content: ElementContent
         var element: Element
 
-        func measure(in constraint: SizeConstraint) -> CGSize {
-            return content.measure(in: constraint)
-        }
+    }
+
+}
 
+private struct EnvironmentAdaptingStorage: ContentStorage {
+    let childCount = 1
+
+    var layout: SingleChildLayout
+
+    /// During measurement or layout, the environment adapter will be applied
+    /// to the environment before passing it
+    ///
+    var adapter: (inout Environment) -> Void
+
+    var child: Element
+
+    func performLayout(
+        attributes: LayoutAttributes,
+        environment: Environment)
+        -> [(identifier: ElementIdentifier, node: LayoutResultNode)]
+    {
+        let environment = adapted(environment: environment)
+
+        let childAttributes = layout.layout(
+            size: attributes.bounds.size,
+            child: child.content.measurable(in: environment))
+
+        let identifier = ElementIdentifier(key: nil, index: 0)
+
+        let node = LayoutResultNode(
+            element: child,
+            layoutAttributes: childAttributes,
+            content: child.content,
+            environment: environment)
+
+        return [(identifier, node)]
     }
 
+    func measure(in constraint: SizeConstraint, environment: Environment) -> CGSize {
```

Comment:
Unrelated to this PR, but I wonder if it's worth merging layout and measurement... Since we can derive measurement from layout, seems like the system could be made more efficient by merging them.

Reply:
I've thought about that a bunch actually, and I think it would be great! But I haven't had time to experiment and see if there are any gotchas.